collectors/fda_dap/spider.py

```diff
@@ -0,0 +1,266 @@
+            raise Exception(msg)
+
+    def parse_drug_details(self, response):
+        '''
+        We're getting closer! We're already on the drug details page, now we
+        just need to look for the Approval History link (if it exists). The
+        state is still kept in the cookies, so we need to make sure we're using
+        the same CookieJar as the one used on the request here.
+        '''
+        url_candidates = response.css('#user_provided li > a::attr(href)').extract()
+        for url in url_candidates:
+            if 'Search.Label_ApprovalHistory' in url:
+                meta = {
+                    'cookiejar': response.meta.get('cookiejar'),
+                }
+                return Request(
+                    urlparse.urljoin(response.url, url),
+                    meta=meta,
+                    dont_filter=True,
+                    callback=self.parse_approval_history
+                )
+
+    def parse_drug_overview(self, response):
+        '''
+        This page contains multiple applications for a single drug. We need to
+        go to the Drug Details page of each one of those. When we click on a
+        link here, we're redirected to a page which sets the state (in the
+        cookies) to the drug's details, then redirecting us to the drug details
+        page. To avoid the requests messing up with each other, we need a clean
+        set of cookies for each one.
+        '''
+        urls = response.css('.product_table a::attr(href)').extract()
+        for url in urls:
+            meta = {
+                'cookiejar': random.random(),
+            }
+            yield Request(
+                urlparse.urljoin(response.url, url),
+                meta=meta,
+                dont_filter=True,
+                callback=self.parse_drug_details_or_overview
+            )
+
+    def _response_cookies_to_dict(self, response):
+        cookies = CookieJar().make_cookies(response, response.request)
+
+        def _cookie_to_dict(cookie):
+            return {
+                'name': cookie.name,
+                'value': cookie.value,
+                'domain': cookie.domain,
+                'path': cookie.path,
+            }
+
+        return [_cookie_to_dict(cookie) for cookie in cookies]
+
+    def parse_approval_history(self, response):
+        _extract_text = lambda selector: selector.xpath('text()') \
+                                                 .extract_first() \
+                                                 .strip()
+
+        def _parse_links(selector):
+            names = [_clean_document_name(name)
+                     for name in selector.css('a::text').extract()]
+            urls = [url.strip()
+                    for url in selector.css('a::attr(href)').extract()]
+            msg = 'The number of document names and URLs is different (%d != %d)' % (len(names), len(urls))
+            assert len(names) == len(urls), msg
+
+            return [{'name': foo[0], 'urls': [foo[1]]}
```

Comment:
Based on old discussion about style =)
I need significant amount of time to get what this statement does) So targeting an average coder like me I'm trying to use those old school things like loops etc. It's very controversial strategy in you know this be pythonic world but could be good in multi-developer multi-skills dev team.